pkg/cmd/codespace/ssh.go

```diff
@@ -298,6 +310,103 @@ func checkAndUpdateOldKeyPair(sshContext ssh.Context) *ssh.KeyPair {
 	return nil
 }
 
+// hasUploadedValidCustomKey checks which private keys are used in ssh configration and
+// compares them to uploaded public keys to see if there is a match
+func hasUploadedPublicKeyForConfig(
+	ctx context.Context,
+	sshContext ssh.Context,
+	apiClient apiClient,
+) bool {
+	configuredPrivateKeyPaths, err := getConfiguredPrivateKeys(ctx)
+	if err != nil {
+		return false
+	}
+
+	configuredPublicKeys := []string{}
+	for _, privateKeyPath := range configuredPrivateKeyPaths {
+		if _, err := os.Stat(privateKeyPath); err != nil {
+			// The default configuration includes standard keys like id_rsa,
+			// id_ed25519, etc, but these may not actually exist
+			continue
+		}
+
+		configuredPublicKey, err := sshContext.GetPublicKeyFromPrivateKey(privateKeyPath)
+		if err != nil {
+			continue
+		}
+
+		configuredPublicKeys = append(configuredPublicKeys, configuredPublicKey)
+	}
+
+	if len(configuredPublicKeys) == 0 {
+		// There are no local private keys which ssh would use
+		return false
+	}
+
+	user, err := apiClient.GetUser(ctx)
+	if err != nil {
+		return false
+	}
+
+	uploadedPublicKeys, err := apiClient.AuthorizedKeys(ctx, user.Login)
+	if err != nil {
+		return false
+	}
+
+	if len(uploadedPublicKeys) == 0 {
+		return false
+	}
+
+	for _, configuredPublicKey := range configuredPublicKeys {
+		for _, uploadedPublicKey := range uploadedPublicKeys {
+			if configuredPublicKey == uploadedPublicKey {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// getConfiguredPrivateKeys reads the effective configuration for a localhost
+// connection and returns all private keys which would be tried for authentication
+func getConfiguredPrivateKeys(ctx context.Context) ([]string, error) {
+	sshExe, err := safeexec.LookPath("ssh")
+	if err != nil {
+		return nil, fmt.Errorf("could not find ssh executable: %w", err)
+	}
+
+	// The -G option tells ssh to output the effective config for the given host, but not connect
+	// The target host is always localhost because we skip this check if --profile was set.
+	sshGCmd := exec.CommandContext(ctx, sshExe, "-G", "localhost")
```

Comment:
The issue I had with this was that we can't assume that files in `~/.ssh` which don't end in `.pub` are private keys. For example, there's `~/.ssh/config` or `~/.ssh/codespaces` which are just config files, but because they have no file extension it's hard to programmatically differentiate them from private keys. Reading the effective config helps in 2 ways:

- it tells me exactly which files I need to try to read as private keys
- it also tells me which private keys I should ignore

The second point is very important - say that I have a public key uploaded to GH, and have the private key locally in `~/.ssh`, BUT that that private key isn't in my default `~/.ssh/config`. If the logic here blindly matched that public/private key then I would still run into the impossible login issue because my configuration didn't tell `ssh` to use that private key. Using `-G` here let's the code know that that private key wouldn't actually apply and thus avoid this false-positive match.

Reply:
Ah, very good point about false-positives. Then I'm in favor of the `ssh -G localhost` approach, and we could even modify `ssh.Context` to use that approach too.

Reply:
I'm don't think it makes sense to update the existing public key lookup logic to make use of `-G`. That logic is by `gh ssh-key add` and `gh auth login -p ssh` in which case even though you are working with ssh keys, you're not really in an ssh context. For example, there's no hostname to use for `ssh -G <hostname>` and anyways it would only return you a subset of local keys even though for the command in question ANY public key should be fine.